src/adhan/calculation_methods.ts

```diff
@@ -0,0 +1,381 @@
+    !location ||
+    !location.latitude ||
+    !location.longitude ||
+    !calculationMethod
+  ) {
+    throw new Error('Required properties are missing from the settings.');
+  }
+
+  const prayerTimeOptions: PrayerTimesOptions = {
+    calculationParameters: CalculationMethods[calculationMethod].get(settings),
+    coordinates: new Coordinates(location.latitude, location.longitude),
+    calculationMethod: CalculationMethods[calculationMethod],
+    midnightMethod: midnightMethod,
+    midnightAdjustment: midnightAdjustment ?? 0,
+  };
+
+  setRounding(prayerTimeOptions, roundingMethod);
+  setAdjustments(prayerTimeOptions, settings);
+  setOverrides(prayerTimeOptions, settings);
+  setHighLatitudeRule(prayerTimeOptions, highLatitudeRule);
+  setMadhab(prayerTimeOptions, asrCalculation);
+  setShafaq(prayerTimeOptions, shafaq);
+  setPolarCircleResolution(prayerTimeOptions, polarResolution);
+
+  return prayerTimeOptions;
+}
+
+function setRounding(
+  prayerTimeOptions: PrayerTimesOptions,
+  roundingMethod?: typeof Rounding[keyof typeof Rounding]
+) {
+  prayerTimeOptions.calculationParameters.rounding =
+    roundingMethod || prayerTimeOptions.calculationParameters.rounding;
+}
+
+function setAdjustments(
+  prayerTimeOptions: PrayerTimesOptions,
+  state: CalculationSettings
+) {
+  prayerTimeOptions.calculationParameters.adjustments = {
+    fajr: state.fajrAdjustment ?? 0,
+    sunrise: state.sunriseAdjustment ?? 0,
+    dhuhr: state.dhuhrAdjustment ?? 0,
+    asr: state.asrAdjustment ?? 0,
+    sunset: state.sunsetAdjustment ?? 0,
+    maghrib: state.maghribAdjustment ?? 0,
+    isha: state.ishaAdjustment ?? 0,
+  };
+}
+
+function setOverrides(
+  prayerTimeOptions: PrayerTimesOptions,
+  state: CalculationSettings
+) {
+  const params = prayerTimeOptions.calculationParameters;
+  params.fajrAngle = state.fajrAngleOverride || params.fajrAngle;
+  params.ishaAngle = state.ishaAngleOverride || params.ishaAngle;
+  params.ishaInterval = state.ishaIntervalOverride || params.ishaInterval;
+  params.maghribAngle = state.maghribAngleOverride || params.maghribAngle;
+}
+
+function setHighLatitudeRule(
+  prayerTimeOptions: PrayerTimesOptions,
+  highLatRuleSetting?: typeof HighLatitudeRule[keyof typeof HighLatitudeRule]
+) {
+  const rule =
+    highLatRuleSetting ??
+    HighLatitudeRule.recommended(prayerTimeOptions.coordinates);
+  switch (rule) {
+    case HighLatitudeRule.MiddleOfTheNight:
+    case HighLatitudeRule.SeventhOfTheNight:
+    case HighLatitudeRule.TwilightAngle:
+      prayerTimeOptions.calculationParameters.highLatitudeRule = rule;
+      break;
+    default:
+      prayerTimeOptions.calculationParameters.highLatitudeRule =
+        HighLatitudeRule.recommended(prayerTimeOptions.coordinates);
+  }
+}
+
+function setMadhab(
+  prayerTimeOptions: PrayerTimesOptions,
+  asrCalcSetting: string
+) {
+  prayerTimeOptions.calculationParameters.madhab =
+    asrCalcSetting === Madhab.Hanafi ? Madhab.Hanafi : Madhab.Shafi;
```

Comment:
good suggestion, unfortunately, TS is complaining about no namespace found even thou I am already using it, most likely it's due to the rest of the code we pass in a string rather than an enum, but this seems to work.   

```ts
function setMadhab(
  prayerTimeOptions: PrayerTimesOptions,
  asrCalcSetting: typeof Madhab[keyof typeof Madhab] = Madhab.Shafi
) {
  prayerTimeOptions.calculationParameters.madhab = asrCalcSetting;
}
```